src/script/interpreter.cpp

```diff
@@ -2051,7 +2051,38 @@ static bool VerifyTaprootSpend(std::vector<valtype> stack,
     if (!IsPayToTaproot(script_pubkey)) {
         return set_error(serror, ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT);
     }
-    // Temporarily return UNKNOWN
+    valtype vch_pubkey =
+        valtype(script_pubkey.begin() + TAPROOT_INTRO_SIZE,
+                script_pubkey.begin() + TAPROOT_SIZE_WITHOUT_STATE);
+
+    if (stack.size() == 0) {
+        return set_error(serror, ScriptError::INVALID_STACK_OPERATION);
+    }
+    if (stack.size() >= 2 && !stack.back().empty() &&
+        stack.back()[0] == TAPROOT_ANNEX_TAG) {
+        return set_error(serror, ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED);
+    }
+    if (stack.size() == 1) {
+        // Spend using single signature instead of executing script
+        valtype &vch_sig = stacktop(-1);
+        bool f_success = false;
+        if (!EvalChecksig(vch_sig, vch_pubkey, script_pubkey.begin(),
```

Comment:
I would almost check for null here just to be sure.

Reply:
```C++
if (!vchSig.size()) {
  return set_error(serror,ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED);
}
...
assert(f_success);
```

Reply:
My only concern with this, is that we don't test something and then end up with the assert in a place where a remote crash can happen.